[PowerPC][AIX] Emit weak_definition symbols as weak externals

aabhinavg1 opened the pull request:
`.weak_definition` is a Mach-O-specific directive and is not supported on XCOFF. Currently, assembling `.weak_definition` for AIX crashes the XCOFF streamer (`report_fatal_error("Not implemented yet.")`).

This patch returns `false` from `MCXCOFFStreamer::emitSymbolAttribute` for `MCSA_WeakDefinition`, matching `MCELFStreamer`, so the parser rejects `.weak_definition` with the same diagnostic as ELF instead of crashing. Mach-O behavior is unchanged.

Test:
- `llvm/test/MC/PowerPC/aix-weak-definition.s`

```sh
$ llvm-mc -triple powerpc-ibm-aix-xcoff test.s
error: unable to emit symbol attribute
```

Testing:
- `check-llvm-mc` (llvm/test/MC): 7765 passed
- `check-llvm-codegen-powerpc`: 1955 passed

Relates to #130269 (the `.weak` directive support itself remains a separate follow-up).

aabhinavg1 committed:
[PowerPC][AIX] Emit weak_definition symbols as weak externals

- On AIX, the assembler does not recognize '.weak_definition'.
- Emit MCSA_WeakDefinition symbols using the standard weak directive (.weak)
  and set storage class to C_WEAKEXT in XCOFF.
- Added test llvm/test/MC/PowerPC/aix-weak-definition.s

aabhinavg1 committed:
Merge branch 'main' into main

aabhinavg1 committed:
Merge branch 'main' into main

tonykuttai commented:
Adding @hubert-reinterpretcast & @daltenty.

aabhinavg1 committed:
Merge branch 'llvm:main' into main

aabhinavg1 committed:
Merge branch 'llvm:main' into main

MaskRay reviewed (commented):
Do you need .weak_definition? ELF doesn't support .weak_definition and we  don't use MCSA_WeakDefinition for ELF. You can use MCSA_Weak for XCOFF

aabhinavg1 commented:
I’ve updated the patch to use **MCSA_Weak** for **XCOFF** instead of **MCSA_WeakDefinition**.

The behaviour remains the same:

.**weak_definition** in inline asm is accepted by LLVM

LLVM emits .weak for XCOFF/AIX targets

The symbol is treated as a weak external (C_WEAKEXT)

Tests have been re-run and pass with the updated implementation.
Please let me know if any further adjustments are needed.

MaskRay reviewed (changes requested):
Then you should just use `MCSA_Weak` instead of `MCSA_WeakDefinition`

aabhinavg1 committed:
[PowerPC][AIX] Emit weak_definition symbols as weak externals

Address review feedback: map .weak_definition to MCSA_Weak in the
generic AsmParser when assembling for XCOFF instead of special-casing
MCSA_WeakDefinition in the streamers, and revert the AIX-specific
streamer changes.

With this mapping, .weak_definition produces the documented .weak
directive in assembler output and a C_WEAKEXT weak external in object
output.

Fixes #130269

aabhinavg1 committed:
[PowerPC][AIX] Reject .weak_definition at parse time

.weak_definition is a Mach-O-specific directive and is not supported on
XCOFF. Reject it at parse time so that assembling .weak_definition for
AIX produces a proper diagnostic instead of mapping it to MCSA_Weak or
crashing the XCOFF streamer.

hubert-reinterpretcast commented on llvm/lib/MC/MCParser/AsmParser.cpp:
I don't see why we wouldn't just do what ELF does, which is to return `false` from `emitSymbolAttribute`.

aabhinavg1 committed:
[PowerPC][AIX] Reject .weak_definition like ELF

.weak_definition is a Mach-O-specific directive and is not supported on
XCOFF. Previously it crashed the XCOFF streamer via report_fatal_error
in emitSymbolAttribute. Return false for MCSA_WeakDefinition like
MCELFStreamer does, so the parser rejects it with the same diagnostic as
ELF.

hubert-reinterpretcast reviewed (approved):
LGTM; thanks! Please wait a day or two to see if other reviewers have comments.
